client_connection/manager.go

```diff
@@ -1,55 +1,99 @@
+	"github.com/mysterium/node/openvpn"
+	vpn_session "github.com/mysterium/node/openvpn/session"
+	"github.com/mysterium/node/server"
+	"github.com/mysterium/node/server/dto"
 )
 
-type State string
+type DialogEstablisherFactory func(identity identity.Identity) communication.DialogEstablisher
 
-const (
-	NOT_CONNECTED = State("NOT_CONNECTED")
-	NEGOTIATING   = State("NEGOTIATING")
-	CONNECTED     = State("CONNECTED")
-)
-
-var (
-	ALREADY_CONNECTED = errors.New("already connected")
-)
+type VpnClientFactory func(vpnSession *vpn_session.VpnSession, session *dto.Session) (openvpn.Client, error)
 
-type ConnectionStatus struct {
-	State     State
-	SessionId string
+type connectionManager struct {
+	//these are passed on creation
+	mysteriumClient          server.Client
+	dialogEstablisherFactory DialogEstablisherFactory
+	vpnClientFactory         VpnClientFactory
+	//these are populated by Connect at runtime
+	dialog    communication.Dialog
+	vpnClient openvpn.Client
+	status    ConnectionStatus
 }
 
-type Manager interface {
-	Connect(identity identity.Identity, NodeKey string) error
-	Status() ConnectionStatus
-	Disconnect() error
-	Wait() error
+func NewManager(mysteriumClient server.Client, dialogEstablisherFactory DialogEstablisherFactory, vpnClientFactory VpnClientFactory) *connectionManager {
+	return &connectionManager{
+		mysteriumClient,
+		dialogEstablisherFactory,
+		vpnClientFactory,
+		nil,
+		nil,
+		statusNotConnected(),
+	}
 }
 
-type fakeManager struct {
-	errorChannel chan error
+func (manager *connectionManager) Connect(identity identity.Identity, NodeKey string) error {
+	manager.status = statusConnecting()
+	session, err := manager.mysteriumClient.SessionCreate(NodeKey)
+	if err != nil {
+		manager.status = statusError(err)
+		return err
+	}
+
+	proposal := session.ServiceProposal
+
+	dialogEstablisher := manager.dialogEstablisherFactory(identity)
+	manager.dialog, err = dialogEstablisher.CreateDialog(proposal.ProviderContacts[0])
+	if err != nil {
+		manager.status = statusError(err)
+		return err
+	}
+
+	vpnSession, err := vpn_session.RequestSessionCreate(manager.dialog, proposal.Id)
+	if err != nil {
+		manager.status = statusError(err)
+		return err
+	}
+
+	manager.vpnClient, err = manager.vpnClientFactory(vpnSession, &session)
+
+	if err := manager.vpnClient.Start(); err != nil {
+		manager.status = statusError(err)
+		return err
+	}
+	manager.status = statusConnected(session.Id)
+	return nil
 }
 
-func NewManager() *fakeManager {
-	return &fakeManager{make(chan error)}
+func (manager *connectionManager) Status() ConnectionStatus {
+	return manager.status
 }
 
-func (nm *fakeManager) Connect(identity identity.Identity, NodeKey string) error {
+func (manager *connectionManager) Disconnect() error {
+	manager.dialog.Close()
+	manager.vpnClient.Stop()
 	return nil
 }
 
-func (nm *fakeManager) Status() ConnectionStatus {
-	return ConnectionStatus{NOT_CONNECTED, ""}
+func (manager *connectionManager) Wait() error {
```

Comment:
We dont need `Wait()` anymore, correct?

Reply:
Well if we are shutting connection manager on client exiting, maybe we would like to wait until it stops gracefully